tools/proofers/my_suggestions.php

```diff
@@ -0,0 +1,760 @@
+            $ELR_round->project_available_state,
+        );
+    }
+
+    // for each project returned from the query, generate a priority field
+    // calculated by the view's weights paired with that field's weights from
+    // the selection criteria. The priority field is used for sorting the rows
+    $now = time();
+    $result = DPDatabase::query($sql);
+    $projects = [];
+    while ($row = mysqli_fetch_assoc($result)) {
+        $row["created"] = hexdec(substr($row["projectid"], 9, 8));
+        $row["round"] = $Round_for_project_state_[$row["state"]]->id;
+
+        $priority = 1;
+        foreach ($view_weights as $field => $weight) {
+            // dynamic fields calculated from project
+            if ($field == "created") {
+                $field_weight = ($now - $row["created"]) / $now;
+            } elseif ($field == "percent_done") {
+                $field_weight = 1 - $row["percent_done"];
+            } else {
+                $normalized_field_value = $row[$field];
+                // only use the primary language for the project when weighting
+                if ($field == "language") {
+                    $normalized_field_value = Project::decode_language($row[$field])[0];
+                }
+
+                // Default field weight needs to be non-zero
+                $field_weight = 0.01;
+
+                // get the field weight from the user weights if one exists,
+                // otherwise try the site weights
+                if(isset($user_weights[$field][$normalized_field_value])) {
+                    $field_weight = $user_weights[$field][$normalized_field_value];
+                } elseif (isset($site_weights[$field][$normalized_field_value])) {
+                    $field_weight = $site_weights[$field][$normalized_field_value];
+                }
+            }
+            $priority *= $weight * $field_weight;
+        }
+        $row["priority"] = $priority;
+
+        $projects[] = $row;
+    }
+    mysqli_free_result($result);
+
+    uasort($projects, "compare_array_by_priority");
+
+    $user = new User($username);
+    $return_projects = [];
+    // loop through the sorted rows checking each one against the ability
+    // for the proofreader to work on it *right now* based on the reserve, etc.
+    foreach (array_reverse($projects) as $project_row) {
+        // only return 20 projects
+        if (count($return_projects) >= 20) {
+            break;
+        }
+
+        // Build the project from the row so we don't have to go back to
+        // the database again for every project. This requires all the necessary
+        // fields used here, and in dependent functions, be populated in the row.
+        $project = new Project($project_row);
+
+        try {
+            $round = $Round_for_project_state_[$project->state];
+            validate_user_against_project_reserve($user, $project, $round);
+            $return_projects[] = $project_row;
+        } catch(Exception $exception) {
```

Comment:
I was originally wondering if any other exceptions can occur that we should not ignore?

Reply:
Right now this can only return `ReservedForNewProofreadersException` but I thought catching any exception was a safer approach here as any exception very likely indicates that validation failed for some reason or another.